---
title: "Practical Exercises Statistics"
author: "Oscar Baeza Francisoud"
date: "28/11/2019"
output:
  html_document:
    toc: true
    toc_float: true
    theme: cerulean
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Practical 1 Statistics
#1.	Convert variables "y" and "gender" to factor variables
#2.	Define the labels for variables "y" and "gender".


We first open the dataset, in this case bladder.1.txt:
```{r chunk2,tidy=TRUE}
rm(list=ls())
setwd("~/UNI/Master/Estadistica/Task 2 practice statistics-20191117")
bladder <- read.csv(file = "bladder.1.txt",header = TRUE,sep = "")
knitr::kable(head(bladder,n = 5),caption = "Bladder DataSet")
```

Then we convert "y" and "gender" to factor, using this code:
```{r chunk3}
bladder$y <- factor(bladder$y, levels = c(0,1),labels = c("control", "case"))
bladder$gender <- factor(bladder$gender, levels = c(1,2), labels = c("male","female"))
knitr::kable(head(bladder,n = 5),caption = "Bladder DataSet with y and gender as factors")

```

#3.	Build a frequency table for "y" that contains, both, the absolute and the relative frequencies. The same for "gender".

First the tables for variable "y":

```{r chunk4}
freq.y <- table(bladder$y)  #absolute frequency
freq.y   
relfreq.y <- prop.table(table(bladder$y))   #relative frequency
relfreq.y   
totfreq.y <- cbind(freq.y,relfreq.y)   #unite both tables
totfreq.y
```

Then the tables for variable "gender":

```{r chunk5}
freq.gen <- table(bladder$gender)  #absolute frequency
freq.gen   
relfreq.gen <- prop.table(table(bladder$gender))   #relative frequency
relfreq.gen   
totfreq.gen <- cbind(freq.gen,relfreq.gen)   #unite both tables
totfreq.gen
```

#4.	Plot gene1 levels as a function of "gender":

```{r chunk6}
plot(bladder$gender, bladder$gene1,xlab="Gender",ylab="Gene1 Expression",
     main="Gene1 Expression Levels as a function of gender")
```

#5.	Test the normality of the expression levels of the 20 genes (use function apply). How many genes are not normally distributed and which are these genes?

We test the normality using the `shapiro.test` function:

```{r chunk7}
pval <- do.call(rbind, lapply(bladder[4:23], function(x) shapiro.test(x)[c("p.value")]))
nonorm <- (pval<0.05) #Selecting those with p.value < 0.05
nonorm <- subset(nonorm, nonorm == "TRUE")
nonorm
```

We select the p-values lower than `0.05`, wich is our confidence interval. For those genes that have this lower p-value, we reject the null hypothesis and they will not be normally distributed. The rest of the genes we cannot reject the null hypothesis. The not normally distributed genes are 2 genes: `gene8`and `gene18`. 

#6.	Test whether mean expression levels of `gene1` and `gene2` are equal. Note: this is a test for equality of means with paired samples. This test is performed by computing the difference of the two variables (gene1-gene2) and testing whether the mean of the difference is equal to zero.

As the note tells us, we first have to compute the difference of the two variables due to them being paired, and then we perform a `t.test` for the equality of two means:

```{r chunk8}
diff <- bladder$gene1-bladder$gene2
t.test(diff,mu = 0)
```

We obtain a p-value of `0.9898`, which is greater than `0.05`. Due to this, we cannot reject the null hypothesis and we assume the mean expression levels of `gene1` and `gene2` are equal.

#7.	Test if the mean expression levels of gene1 are equal between cases and controls.

To do this, we first have to know if the variances are equal or not, because we will do different tests depending on this information. The test for detecting this equality of variances is `var.test`:

```{r chunk9, tidy=TRUE}
var.test(bladder$gene1~bladder$y)
```

The p-value the `var.test`provides us with is `0.7906`,which is greater than our confidence interval and so we cannot reject the null hypothesis and we assume the relation between variances from variables `gene1` and `y` is equal to 1. 

Following the last sentence, we now know which kind of equality test we can perform, a Student's `t.test`:

```{r chunk10}
t.test(bladder$gene1~bladder$y, var.equal=T)
```

The result of this test is a p-value of `0.4615`,which is greater than our confidence interval. With this, we cannot reject the null hypothesis and we assume the difference in means is equal to zero betwenn these two variables.

#8.	Obtain a 95% bootstrap confidence interval for the 20th percentile of gene1: 

```{r chunk11}
library(boot)
bootfun<-function(data,N){
  quant_20 <- NULL
  for (i in 1:N) {
    newsample <- sample(data,length(data),replace = T)
    quant_20 <- c(quant_20, quantile(newsample,probs = 0.2))
  }
return(quant_20)  
}

boot20 <- bootfun(bladder$gene1,10000)
confidn <- c(quantile(boot20,0.025),quantile(boot20,0.975))
confidn
quantile(bladder$gene1,.2)
```

#9.	Contrast using a permutation test whether the 20th percentile of gene1 for male and female are equal:

```{r chunk11}
bootfun_null<-function(data,N){
  data_s <- sample(data)
  quant_20 <- NULL
  for (i in 1:N) {
    newsample <- sample(data_s,length(data_s),replace = T)
    quant_20 <- c(quant_20, quantile(newsample,probs = 0.2))
  }
  return(quant_20)  
}

boot20perm <- bootfun_null(bladder$gene1,10000)
hist(boot20perm)
p.value<-(sum(boot20perm>boot20)+1)/(length(boot20perm)+1)
p.value
```

As the pvalue is bigger than 0.05, we cannot reject the null hypothesis and we know that the 20th percentile of gene1 for male and female are equal.

#10.	Perform a nonparametric test for association of gender and the risk of disease. Provide the OR (change the levels of "gender" if necessary in order that the given OR is larger than 1)

We will perform a `chisq.test` to test this hypothesis:
```{r chunk12}
chisq.test(x = bladder$gender,y = bladder$y)
```
The pvalue is bigger than 0.05, so we cannot reject the null and we know that gender and the risk of disease is not associated.

To provide the Odds Ratio, we will first use the function `oddsratio` from package epitools, and then the more manual way by doing the exponential of the coefficient of the linera model:
```{r chunk13}
install.packages("epitools")
library(epitools)
table_OR <- table(bladder$gender,bladder$y)
table_OR
oddsratio(table_OR)
#The OR is 1.263
#No te sentit ja que no estan associades

gen.dis.ass <- glm(bladder$gender~bladder$y, family=binomial()) 
gen.dis.ass
summary(gen.dis.ass)
exp(summary(gen.dis.ass)$coefficients[2,1])
#We confirm the OR.
```
Even though we have and OR of 1.262 for females vs males, it is caused by other variables, not gender, that are associated with the disease and get confused by gender.

#11.	Explore for possible relationship between methylation and gene expression:
```{r chunk14}
meth.gene.ass <- glm(bladder$methyl~bladder$gene1+bladder$gene2+bladder$gene3
                     +bladder$gene4+bladder$gene5+bladder$gene6+bladder$gene7
                     +bladder$gene8+bladder$gene9+bladder$gene10+bladder$gene11
                     +bladder$gene12+bladder$gene13+bladder$gene14+bladder$gene15
                     +bladder$gene16+bladder$gene17+bladder$gene18+bladder$gene19
                     +bladder$gene20, family=binomial())
summary(meth.gene.ass)
```
We can see on the pvalue colummn of the summary of the linear model between methylation and gene expression that only gene expresssion for gene20 is related with methylation. This is the only significant pvalue, so for this gene we reject the null hypothesis and we can say that methylation influences the expression of gene2. For the rest of genes we cannot reject the null hypothesis and so methylation would do nothing on the gene expression.

#12.	Identify genes that are related to the risk of bladder cancer using a multivariate logistic regression model with stepwise variable selection. Denote the selected model as "best.model". Interpret the obtained model.
```{r chunk15}
dis.gene.ass <- glm(bladder$y~bladder$gene1+bladder$gene2+bladder$gene3
                    +bladder$gene4+bladder$gene5+bladder$gene6+bladder$gene7
                    +bladder$gene8+bladder$gene9+bladder$gene10+bladder$gene11
                    +bladder$gene12+bladder$gene13+bladder$gene14+bladder$gene15
                    +bladder$gene16+bladder$gene17+bladder$gene18+bladder$gene19
                    +bladder$gene20, family=binomial())


bothmodel<-step(dis.gene.ass,direction="both")
summary(bothmodel)

best.model <- glm(bladder$y~bladder$gene3+bladder$gene5+bladder$gene8+
                    bladder$gene17+bladder$gene19,family = binomial())

summary(best.model)
```
```{r chunk15}
install.packages("ROCR")
library(ROCR)
lp<-best.model$linear.predictors
pred <- prediction(-lp, bladder$y)
perf <- performance(pred, "tpr", "fpr" )
plot(perf)
abline(a=0, b= 1)
title("ROC curve for best.model; bladder.1 data")

auc<-slot(performance(pred,"auc"), "y.values")[[1]]
auc
```
```{r chunk16}
K <- 5
n <- nrow(bladder)  #number of individuals
fold <- sample(as.numeric(cut((1:n),breaks = K)))  # random assignment of each individual into one fold 

pred <- NULL # vector of predictions

for(i in 1:K){
  indTest <- which(fold==i)   # Test indices 
  indTrain <- which(fold!=i)  # Train indices
  model.i <- glm(bladder$y~bladder$gene3+bladder$gene5+bladder$gene8+
                   bladder$gene17+bladder$gene19, data=bladder[indTrain,], family=binomial())  # Adjust the model with training data
  pred.i <- predict(model.i, newdata=bladder[indTest, ])   # Predicts test data at step i
  pred[indTest] <- pred.i   # Store predicted values for test data at step i 
}  

prediction <- prediction(pred, bladder$y) 
perf <- performance(prediction, "tpr", "fpr" )
plot(perf,main="Crossvalidation for best.model with k=5")
abline(a=0, b= 1)

auc<-slot(performance(prediction,"auc"), "y.values")[[1]]
auc
```
```{r chunk17}
bladder2 <- bladder[,4:23]
genes <- ifelse(bladder2>=0,"high","low")
colnames(genes) <- paste(rep("levels.gene",20),seq(1,20,1),sep="")
bladder2 <- data.frame(genes)
knitr::kable(head(bladder2,n = 5),caption = "Bladder_2 DataSet")
str(bladder2$levels.gene1)
```
```{r chunk18}
pvalues <- NULL
names <- NULL
positions <- grep("levels.gene",ls())
positions
```
```{r chunk19}
dist1 <- dist(bladder[, c(25:50)],method="euclidian")
diam <- as.matrix(dist1)
dim(diam)
diam[1:10]

plot(hclust(dist1,method="average"))
plot(hclust(dist1,method="ward.D2"))

kmeans2<-kmeans(dist1,2)
plot(dist1, col = kmeans2$cluster)
points(kmeans2$centers,col = 1:2, pch = 8, cex=2)
```
```{r chunk20}
data16<-t(bladder[, c(25:50)])
distind <- dist(data16,method="euclidian")
distind <- as.matrix(distind)
dim(distind)

plot(hclust(dist(data16,method="euclidian"),method="ward.D2"))
plot(hclust(dist(data16,method="euclidian"),method="average"))

kmeans3<-kmeans(data16,2)

plot(data16, col = kmeans3$cluster)
points(kmeans3$centers,col = 1:2, pch = 8, cex=2)
```